internal/controllers/machineset/machineset_preflight_test.go

```diff
@@ -0,0 +1,472 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machineset
+
+import (
+	"testing"
+
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	utilfeature "k8s.io/component-base/featuregate/testing"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	"sigs.k8s.io/cluster-api/feature"
+	"sigs.k8s.io/cluster-api/internal/contract"
+	"sigs.k8s.io/cluster-api/internal/test/builder"
+)
+
+func TestMachineSetReconciler_runPreflightChecks(t *testing.T) {
+	ns := "ns1"
+
+	controlPlaneWithNoVersion := builder.ControlPlane(ns, "cp1").Build()
+
+	controlPlaneProvisioning := builder.ControlPlane(ns, "cp1").
+		WithVersion("v1.25.6").Build()
+
+	controlPlaneUpgrading := builder.ControlPlane(ns, "cp1").
+		WithVersion("v1.26.2").
+		WithStatusFields(map[string]interface{}{
+			"status.version": "v1.25.2",
+		}).
+		Build()
+
+	controlPlaneStable := builder.ControlPlane(ns, "cp1").
+		WithVersion("v1.26.2").
+		WithStatusFields(map[string]interface{}{
+			"status.version": "v1.26.2",
+		}).
+		Build()
+
+	t.Run("should run preflight checks if the feature gate is enabled", func(t *testing.T) {
+		defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.MachineSetPreflightChecks, true)()
+
+		tests := []struct {
+			name         string
+			cluster      *clusterv1.Cluster
+			controlPlane *unstructured.Unstructured
+			machineSet   *clusterv1.MachineSet
+			wantPass     bool
```

Comment:
Would be good to also test the error cases

Reply:
I have added test cases to cover the important error cases. For the other cases, example: `IsProvisioning`, it looks like it would basically become equivalent to testing those functions and the current tests don't seem like the right place to test them.

Reply:
Sounds good!